src/api/io.controller.js

```diff
@@ -0,0 +1,137 @@
+/*
+  # This Source Code Form is subject to the terms of the Mozilla Public
+  # License, v. 2.0. If a copy of the MPL was not distributed with this
+  # file, You can obtain one at https://mozilla.org/MPL/2.0/.
+*/
+
+'use strict';
+
+// The permissions controller, allows/disallows usage of the module
+const cPerms = require('./permissions.js');
+
+// Contextual pointers provided by the index.js process
+let ptrWALLET;
+let ptrGetMsgFromTx;
+let ptrIsFullnode;
+let strModule;
+let COIN;
+
+function init(context) {
+    ptrWALLET = context.WALLET;
+    ptrGetMsgFromTx = context.getMsgFromTx;
+    ptrIsFullnode = context.isFullnode;
+    // Static Non-Pointer (native value)
+    strModule = context.strModule;
+    COIN = context.COIN;
+    // Initialize permissions controller
+    return cPerms.init({ 'DB': context.DB, 'strModule': strModule });
+}
+
+async function readTx(req, res) {
+    if (!cPerms.isModuleAllowed(strModule)) {
+        return disabledError(res);
+    }
+    if (!ptrIsFullnode()) {
+        return fullnodeError(res);
+    }
+    if (!req.params.txid || req.params.txid.length !== 64) {
+        return res.json({
+            'error': "You must specify a 'txid' param!"
+        });
+    }
+    if (!req.params.format || !req.params.format.length) {
+        return res.json({
+            'error': "You must specify an 'format' param!"
+        });
+    }
+    const strFormat = req.params.format;
+    const strTxid = req.params.txid;
+    // Check the encoding format is valid
+    if (!Buffer.isEncoding(strFormat)) {
+        return res.json({
+            'error': 'Encoding format (' + strFormat + ') is invalid'
+        });
+    }
+    res.json(await ptrGetMsgFromTx(strTxid, strFormat));
+}
+
+async function writeTx(req, res) {
+    if (!cPerms.isModuleAllowed(strModule)) {
+        return disabledError(res);
+    }
+    if (!ptrIsFullnode()) {
+        return fullnodeError(res);
+    }
+    if (!req.params.address || req.params.address.length !== 34) {
+        return res.status(400).send('Missing "address" parameter!');
+    }
+    if (!req.params.data) {
+        return res.status(400).send('Missing "data" parameter!');
+    }
```

Comment:
Is there a length limit for data? If so, validate

Reply:
Added serialized byte-buffer validation in the latest commit, also tested deeply with 'maximized' bytes to ensure it's precise.